File: src/main/cpp/subsystems/climber/ClimberExtend.cpp
#include "subsystems/climber/ClimberExtend.h"
#include "Constants.h"


ClimberExtend::ClimberExtend() {
    // Setup current limits, idle modes, and encoder factors
    winch.SetSmartCurrentLimit(
            CONSTANTS::CLIMBER::EXTEND::HARD_CURRENT_LIMIT.to<int>());
    winch.SetSecondaryCurrentLimit(
            CONSTANTS::CLIMBER::EXTEND::SOFT_CURRENT_LIMIT.to<double>());
    winch.GetEncoder().SetPositionConversionFactor(
            CONSTANTS::CLIMBER::EXTEND::GEARBOX_RATIO.to<double>());
            // The encoder is now in units of drum turns

    // Set up PID controller
    pid.SetP(CONSTANTS::CLIMBER::EXTEND::PID::P);
    pid.SetI(CONSTANTS::CLIMBER::EXTEND::PID::I);
    pid.SetD(CONSTANTS::CLIMBER::EXTEND::PID::D);
    pid.SetFF(CONSTANTS::CLIMBER::EXTEND::PID::FF);
    pid.SetIZone(CONSTANTS::CLIMBER::EXTEND::PID::I_ZONE);
    pid.SetOutputRange(CONSTANTS::CLIMBER::EXTEND::PID::MIN,
            CONSTANTS::CLIMBER::EXTEND::PID::MAX);
}

units::meter_t ClimberExtend::Get() {
    return distance;
}

void ClimberExtend::Set(units::meter_t distance) {
    // Save the distance
    ClimberExtend::distance = distance;

    // Update the PID controller
    pid.SetReference(GetTurns(distance).to<double>(),
                rev::CANSparkMaxLowLevel::ControlType::kPosition);
}

units::turn_t ClimberExtend::GetActual() {
    return (units::turn_t) encoder.GetPosition();
}

units::turn_t ClimberExtend::GetTurns(units::meter_t distance) {
    // Note: This only accounts for a single circumference,
    //  not the changing circumference of the pulley system
    return distance * 1_tr / CONSTANTS::CLIMBER::EXTEND::PULLEY_CIRCUMFERENCE;
}

units::meter_t ClimberExtend::GetDistance(units::turn_t turns) {
    // Note: This only accounts for a single circumference,
    //  not the changing circumference of the pulley system
    return turns * CONSTANTS::CLIMBER::EXTEND::PULLEY_CIRCUMFERENCE / 1_tr;
}

File: src/main/java/frc/bionicpolars/subsystems/climber/ClimberExtend.java
package frc.bionicpolars.subsystems.climber;

import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.SparkMaxPIDController;

import frc.bionicpolars.Constants;

public class ClimberExtend {
    private static final double GEARBOX_RATIO = 100.0,
            PULLEY_DIAMETER = 1.7887,
            PULLEY_CIRCUMFERENCE = Math.PI * PULLEY_DIAMETER,
            P = 2.0,
            I = 0.0,
            D = 0.0;
    private CANSparkMax motor = new CANSparkMax(
            Constants.CLIMBER_EXTENSION_ID,
            CANSparkMax.MotorType.kBrushless);
    private RelativeEncoder encoder = motor.getEncoder();
    private SparkMaxPIDController pid = motor.getPIDController();
    private double distance = 0.0;


    private double getTurns(double distance) {
        // Note: This only accounts for a single circumference,
        //  not the changing circumference of the pulley system
        return distance / PULLEY_CIRCUMFERENCE;
    }

    private double getDistance(double turns) {
        // Note: This only accounts for a single circumference,
        //  not the changing circumference of the pulley system
        return turns * PULLEY_CIRCUMFERENCE;
    }

    public ClimberExtend() {
        encoder.setPositionConversionFactor(GEARBOX_RATIO);
        pid.setP(P);
        pid.setI(I);
        pid.setD(D);
        pid.setOutputRange(-1.0, 1.0);
    }

    public double get() {
        return distance;
    }

    public void set(double distance) {
        this.distance = distance;
        pid.setReference(getTurns(distance), CANSparkMax.ControlType.kPosition);
    }

    public double getActual() {
        return getDistance(encoder.getPosition());
    }

    public void zero() {
        encoder.setPosition(0.0);
    }
}
